### PR TITLE
ci: enable building files caching

### PR DESCRIPTION
**Pull request by Ink-33**
### Ⅰ. Describe what this PR did
Cache go dependencies and compile caches, submodules and prebuild artifacts. The workflow can be completed within **ten minutes** after the cache is ready. 

### Ⅱ. Does this pull request fix one issue?

no

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
no

### Ⅳ. Describe how to verify it
When the cache is ready, a significant reduction in workflow runtime can be observed.

### Ⅴ. Special notes for reviews
Two tricks are used here to accomplish expectations:
1. `.git/modules` is cached to ensure that `git submodule update -init` runs correctly.
2. `git stash` will be run after the cache is pulled to avoid patches files being overwritten by the cache.

These two points require additional testing to avoid unexpected errors. I will reply here with the test results after a brief test.

### CODE REVIEW

**Comment by Ink-33** [comment]
updated upstream

File: .github/workflows/build-and-test.yaml [hunk 1]
```diff
@@ -67,14 +111,39 @@ jobs:
...
+            istio
+            external
+            .git/modules
+        key: ${{ runner.os }}-submodules-${{ github.run_id }}
+        restore-keys: ${{ runner.os }}-submodules
+          
+    - run: git stash # restore patch
+
     - name: "Run Ingress Conformance Tests"
       run: GOPROXY="https://proxy.golang.org,direct" make ingress-conformance-test
```
Comment (Xunzhuo) [inline .github/workflows/build-and-test.yaml:145]: Can we add this into ingress-conformance-test ?

**Comment by Xunzhuo** [comment]
I want to see the difference before and after enabling cache. It takes about 22 mins before caching. Let me rerun build and test workflow and see how long it takes now.

**Review by Xunzhuo — APPROVED** [review]

**Comment by Xunzhuo** [comment]
22 min -> 15 min : )

**Pull request merged**